services/worker/src/worker/main.py

```diff
@@ -1,53 +1,90 @@
+        banner = self.app_config.worker.state_path
+        if not banner:
+            raise ValueError("Failed to create the executor because WORKER_STATE_PATH is missing.")
+        start_worker_loop_command = [
+            sys.executable,
+            START_WORKER_LOOP_PATH,
+            "--print-worker-state-path",
+        ]
+        return OutputExecutor(start_worker_loop_command, banner, timeout=10)
+
+    def start(self) -> None:
+        worker_loop_executor = self._create_worker_loop_executor()
+        worker_loop_executor.start()  # blocking until the banner is printed
+        logging.info("Starting heartbeat.")
+        while worker_loop_executor.running():
+            self.heartbeat()
+            time.sleep(self.app_config.worker.heartbeat_time_interval_seconds)
+        worker_loop_executor.stop()
+
+    def get_state(self) -> WorkerState:
+        worker_state_path = self.app_config.worker.state_path
+        if not worker_state_path:
+            raise ValueError("Failed to get worker state because WORKER_STATE_PATH is missing.")
+        if os.path.exists(worker_state_path):
+            with FileLock(worker_state_path + ".lock"):
+                try:
+                    with open(worker_state_path, "r") as worker_state_f:
+                        worker_state = json.load(worker_state_f)
+                        return WorkerState(current_job_info=worker_state.get("current_job_info"))
+                except json.JSONDecodeError:
+                    return WorkerState(current_job_info=None)
+        else:
+            return WorkerState(current_job_info=None)
+
+    def get_current_job(self) -> Optional[Job]:
+        worker_state = self.get_state()
+        if worker_state["current_job_info"]:
+            job = Job.objects.with_id(worker_state["current_job_info"]["job_id"])  # type: ignore
+            if job and isinstance(job, Job) and job.status == Status.STARTED:
+                return job
+        return None
+
+    def heartbeat(self) -> None:
+        current_job = self.get_current_job()
+        if current_job:
+            current_job.update(last_heartbeat=get_datetime())
+
 
 if __name__ == "__main__":
-    app_config = AppConfig.from_env()
-
-    init_logging(log_level=app_config.common.log_level)
-    # ^ set first to have logs as soon as possible
-    assets_directory = init_assets_dir(directory=app_config.assets.storage_directory)
-
-    processing_graph = ProcessingGraph(app_config.processing_graph.specification)
-
-    with (
-        LibrariesResource(
-            hf_endpoint=app_config.common.hf_endpoint,
-            init_hf_datasets_cache=app_config.datasets_based.hf_datasets_cache,
-            numba_path=app_config.numba.path,
-        ) as libraries_resource,
-        CacheMongoResource(
-            database=app_config.cache.mongo_database, host=app_config.cache.mongo_url
-        ) as cache_resource,
-        QueueMongoResource(
+    with tempfile.TemporaryDirectory() as tmp_dir:
```

Comment:
Are we sure this cannot raise? Depending on how we deploy, maybe we have no access to the directory, and it cannot be created/deleted (see [/chart](https://github.com/huggingface/datasets-server/tree/main/chart)). If we have `WORKER_STATE_FILE_NAME` in the env, we don't need to create a tmp directory.

Reply:
It's supposed to work - but feel free to connect to a worker and try it out x)

Otherwise we can define a dedicated directory (and we can use the env var `TMPDIR` to tell python to use it)

Reply:
OK, let's it like this for now.